### DISCUSSION TITLE
Assign to variable from commented-out magic command

### DISCUSSION THREAD

**Discussion by janosh** [microsoft/pylance-release#8183] [Enhancement]
[IPython allows](https://ipython.org/ipython-doc/3/interactive/reference.html#manual-capture-of-command-output-and-magic-output) expanding Python variables into magic shell commands and also assigning the command's stdout to a variable:

```py
# %%
num = 42
stdout = !echo {num} # raises: Pylance: Expected expression
print(stdout)
>>> ['42']
```

But it seems there's no way to do this without syntax error.

https://github.com/microsoft/vscode-jupyter/issues/3263 introduced `"jupyter.magicCommandsAsComments": true` which allows to do:

```py
slurm_cmd = """
sbatch --partition icelake --account=SL3-CPU --time 1 --nodes 1 --wrap "python -c 'print(42)'"
"""
slurm_msg = ""
#! slurm_msg = {slurm_cmd}

print(f"{slurm_msg=} at {datetime.now():%Y-%m-%d %H-%M}")
>>> slurm_msg='' at 2022-04-24 20-04
```

The shell command runs as expected but the variable assignment is lost. Any chance `magicCommandsAsComments` could get a feature upgrade?

**Comment by DonJayamanne** [comment]
@janosh looks like there are two issues here.
One is the syntax error and the other is variable value being lost. Is that correct?

**Comment by DonJayamanne** [comment]
Transferring to lsp-middle-ware to ensure statements such as `stdout = !echo {num}` are not flagged as errors.

**Comment by janosh** [comment]
@DonJayamanne That's correct, 2 issues here, though the syntax error can be worked around as in the 2nd code snippet with this clunky construction.

```py
slurm_msg = ""
#! slurm_msg = {slurm_cmd}
```

So the only real issue is the variable assignment being lost. But if there's a proper solution to the syntax error (maybe marking it as acceptable code in `pylance`), this could fix the lost assignment as well and in a cleaner way.

**Comment by rchiodo** [comment]
The middleware layer is being deprecated (and well never handled this sort of thing anyway). Might be something to do in pylance itself.

**Comment by heejaechang** [comment]
there seems an issue with ipython mode

**Comment by hyacz** [comment]
I ran into a similar problem

<img width="667" alt="image" src="https://user-images.githubusercontent.com/9551197/201811742-5f77b841-b421-4ad3-94a1-ae5f0d13d325.png">

**Comment by rubensa** [comment]
Might be related to: https://github.com/microsoft/pylance-release/discussions/4935 (generated from: https://github.com/microsoft/pylance-release/issues/4934)

**Comment by heejaechang** [comment]
Repro workspace for investigation:
https://github.com/heejaechang/pylance-repro/tree/main/pylance-issues/3114

**Comment by rchiodo** [comment]
Transferring to discussions for up votes. Looks like it's not a very common request though.
